{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "iiQTHTxHOJCL"
      },
      "source": [
        "# Artificial Intellegence\n",
        "### Solving n-Queen Problem With Genetic Algorithm \n",
        "\n",
        "In this exam you have to implement the genetic algorithm to solve the n-queen problem. Please carefully read the following dicription and write your code where __None__ is written.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8nCR3ODQOJCO"
      },
      "source": [
        "#### Introduction to GA\n",
        "Genetic Algorithm (GA) is a search-based optimization technique based on the principles of Genetics and Natural Selection. It is frequently used to find optimal or near-optimal solutions to difficult problems which otherwise would take a lifetime to solve. It is frequently used to solve optimization problems, in research, and in machine learning.\n",
        " 1. Genetic algorithms (GA) is a variant of stochastic beam search.\n",
        " 2. Successor states are generated by combining two parents rather by modifying a single state. The process to generate new individauls form the parents is called crossover.\n",
        " 3. The process is inspired by natural selection.\n",
        " 4. Starts with $k$ randomly generated states, called population. Each state is an individual.\n",
        " 5. An individual is usually represented by a string of 0’s and 1’s, or digits, or a finite set called chromosome.\n",
        " 6. The objective function is called fitness function: better states have high values of fitness function.\n",
        " 7. Pairs of individuals are selected at random for reproduction w.r.t. some probabilities.\n",
        " 8. A crossover point is chosen randomly in the string.\n",
        " 9. Offspring are created by crossing the parents at the crossover point.\n",
        " 10. Each element in the string is also subject to some mutation with a small probability.\n",
        " <img src=\"algo.png\">\n",
        "<caption><center> <u> <font color='purple'> **A pseudocode of genetic algorithm** </u></center></caption>\n",
        " \n",
        " \n",
        "## 8-queens Problem\n",
        "\n",
        "8-queens is a classic computer science problem. To find possible arrangements of 8-queens on a standard $8\\times 8 $ chessboard such that no queens end up in an attacking configuration. Now, if one knows the basics of chess, one can say that a queen can travel either horizontally, vertically, or diagonally. Hence, for 2 or more queens to be in an attacking state, they have to lie either horizontally or vertically or diagonally in-line with the other queen. Figure 1 illustrates that what all are the attacking positions with respect to a queen. As seen, the attacking positions are those where in the queen can move either horizontally, vertically, or diagonally along the chessboard.” OK ” resembles the chessboard square which is considered a safe and non-attacking position. Whereas the one marked with an ” X ” is an attacking placement. Figure 2, shows one of the possible arrangements that serve as a solution to the 8 queens problem.\n",
        "\n",
        "<img src=\"queen_attacking.png\" style=\"width:300px;height:300px;\">\n",
        "<caption><center> <u> <font color='purple'> **Figure 1** </u></center></caption>\n",
        "\n",
        "<img src=\"solution.jpg\" style=\"width:300px;height:300px;\">\n",
        "<caption><center> <u> <font color='purple'> **Figure 2** </u></center></caption>\n",
        "\n",
        "Figure 3 shows one iteration of genetic algorithm for 8-queen problem \n",
        "<img src=\"ga.png\">\n",
        "<caption><center> <u> <font color='purple'> **Figure 3** </u></center></caption>\n",
        "\n",
        "- **Chromosome representation:** In the 8-queen problem, an individual (chromosome) can be represented by a string of digits 1 to 8, that represents the position of the 8 queens in the 8 columns. **Illustration:** a chromosome for the solution (Figure 2) can be represented as a list [3,1,7,5,8,2,4,6]\n",
        "\n",
        "- **Fitness Function:** The number of non-attacking queens in a given configration will be considered as the fitness.Thus the n-queen problem will be treated as a maximization problem. The configration which has highest fintenss value will be the best solution and vice-versa. The fitness ,$f$, can be given as: \n",
        " \n",
        " $f=max_{attacking}-current_{attacking}$\n",
        " \n",
        "  The $current_{attacking}$ represents the number attacking pairs of queens in the given configration (chromosome) while the $max_{attacking}$ represents the maximum possible pairs of attaccking queens. For $n$-queens the $max_{attacking}$ can be given as:\n",
        " \n",
        " $max_{attacking}=\\frac{n(n-1)}{2}$\n",
        " \n",
        "As we mentioned earlier that an individual can be represented as a list of digits. Let c=[3,2,1,3] represents an individual (chromosome). Generally, a location of a particular queen can be given as $(i,j)$, where $i$ is the index and $j$ is the value at that index.  For example, in the given chromosome 'c' the location of first queen can be represented as (0,3). We can compute the $current_{attacking}$ in a very simple way. Consider two queens at location (p,r) and (s,t) respectively. The queens will be considered attacking if any of the two conditons hold. 1) $r=t$ or .2) $|p-s|=|r-t|$\n",
        " \n",
        "The follwing function implement the above fitness function:\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bUclsLUVOJCS",
        "outputId": "f006c990-c680-471a-80a5-25a750c112e4"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "The number of non-attacking queens 3.0\n"
          ]
        }
      ],
      "source": [
        "def fitness(c): # c is the chromosome \n",
        "    n=len(c)# n repesents the number of queens\n",
        "    current_attacking=0\n",
        "    #Code here to compute the number current_attacking  queens\n",
        "    None\n",
        "    f=max_attacking-current_attacking            \n",
        "    return f\n",
        "\n",
        "#Testing The fitenss function. \n",
        "c=[0,3,0,1]\n",
        "#c=[3,1,7,5,8,2,4,6]\n",
        "print('The number of non-attacking queens',fitness(c))\n",
        "\n",
        "# for the above chromosome 'c' your yor fitness function should print \"The number of non-attacking queens=3.0\""
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "F0uLleAROJCe"
      },
      "source": [
        "### Initial population generation\n",
        "As we discussed in the class that genetic algorithm (GA) is a population based algorithm. GA maintain a set of solutions rather than one solution. Normally the initial population is randomly generated. Let population size, $p_{size}=10$. It means that the resulting population $p$ will contain 10 individauls (chromosomes). For random generation of population we  need a python function $random.randint$. The following example illustrates that how to generate random integers between 1 and 100."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "iCosTp85OJCf",
        "outputId": "e3864929-cbb9-4589-ab81-623f63d4485e"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "2\n",
            "4\n",
            "3\n",
            "3\n",
            "2\n"
          ]
        }
      ],
      "source": [
        "import random as rnd\n",
        "for x in range(5):\n",
        "  print(rnd.randint(1,5))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RSO6W1QoOJCg"
      },
      "source": [
        "In our implementation a population $p$ will be a list of list. Let $p=[[c1],[c2]]$ be a population containing two chromosomes. Let $c1=[0,3,0,1]$ and $c2=[1,1,1,1]$ then the population can be given as $p=[[0,3,0,1],[1,1,1,1]]$. In the following function you have to generate an initital population. You need to pass population size, $p_{size}$, $chromosome_length$, $gene_{min}$ and $gene_{max}$ to the function and it will return a valid population $p$. The  $chromosome_{length}$ means the number of genes in each chromosome. In the given example the chromosome $c1$ and $c2$ contain four genes, thus the $chromosome_{length}=4$. The $gene_{min}$ and $gene_{max}$ are the minimum and maximum values that a particular gene can have. For four queen the $gene_{min}$ and $gene_{max }$ may be 0 and 3 or 1 and 4 respectively. "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wa1q_2vSOJCh",
        "outputId": "d73d0f5b-1909-45b5-9183-ee2a9cb42a73"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[[2, 2, 1, 2], [0, 2, 2, 0], [3, 3, 1, 2], [0, 2, 1, 2], [1, 3, 2, 3]]\n"
          ]
        }
      ],
      "source": [
        "def inital_population(p_size, chromosome_length, gene_min, gene_max):\n",
        "    p=[] #initialize the population\n",
        "    for i in range(p_size):\n",
        "        c=[]\n",
        "        for j in range(chromosome_length):\n",
        "            c.append(rnd.randint(gene_min,gene_max))\n",
        "        p.append(c)\n",
        "    return p\n",
        "\n",
        "#testing inital_population\n",
        "p=inital_population(5,4,0,3)\n",
        "print(p)\n",
        "# For the given setup the output should be of the following form\n",
        "#[[2, 3, 3, 0], [1, 2, 0, 2], [2, 1, 3, 2], [3, 1, 0, 3], [2, 1, 1, 0]]\n",
        "        "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wS1CIE2iOJCi"
      },
      "source": [
        "We need to chanege the repesentation of population form $p=[[c1]],[c2]]$ to $p=[[fitness-value-of-c1,[c1]],[fitness-value-of-c2,[c2]]]$  in order to include the fintess of each chromosome. In the follwing code you have to convert the above population $p=[[2, 3, 3, 0], [1, 2, 0, 2], [2, 1, 3, 2], [3, 1, 0, 3], [2, 1, 1, 0]]$ to the form $p=[[4.0, [2, 3, 3, 0]], [4.0, [1, 2, 0, 2]], [3.0, [2, 1, 3, 2]], [3.0, [3, 1, 0, 3]], [3.0, [2, 1, 1, 0]]]$"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_r4PQdESOJCj",
        "outputId": "a3077dd3-ad28-4563-b8bd-edd4ffd40df0"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[[1.0, [2, 2, 1, 2]], [2.0, [0, 2, 2, 0]], [3.0, [3, 3, 1, 2]], [3.0, [0, 2, 1, 2]], [3.0, [1, 3, 2, 3]]]\n"
          ]
        }
      ],
      "source": [
        "p_f=[] # population with fitness\n",
        "for c in p:\n",
        "    p_f.append([fitness(c),c])\n",
        "#Replace the old population p with P_f\n",
        "p=p_f\n",
        "print(p)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "epodH4XZOJCm"
      },
      "source": [
        "### Selection\n",
        "The basic part of the selection process is to stochastically select from one generation to create the basis of the next generation. The requirement is that the fittest individuals have a greater chance of survival than weaker ones. This replicates nature in that fitter individuals will tend to have a better probability of survival and will go forward to form the mating pool for the next generation. Weaker individuals are not without a chance. In nature such individuals may have genetic coding that may prove useful to future generations.  The following table lists a sample population of 4 individuals as explained in Figure 3.\n",
        "\n",
        "\\begin{array}{|r|r|} \\hline\\\\\n",
        "Chromosome & Fitness & \\% of Total \\\\ \\hline\n",
        "24478552&24 &31 \\\\ \\hline\n",
        "32752411&23 &29\\\\ \\hline\n",
        "24415124&20 &26 \\\\ \\hline\n",
        "32543213&11 &14 \\\\ \\hline\n",
        "\\end{array}\n",
        "\n",
        "The fitness values are computed using the fitness function (number of non-attacking pairs of queen) given above. We can see from the table (column Fitness) that individual No. 1 is the fittest and No. 4 is the weakest. Summing these fitness values we can apportion a percentage total of fitness. This gives the strongest individual a value of 31% and the weakest 14%.\n",
        "\n",
        "These percentage fitness values can then be used to configure the roulette wheel. The number of times the roulette wheel is spun is equal to size of the population. As can be seen from the way the wheel is now divided, each time the wheel stops this gives the fitter individuals the greatest chance of being selected for the next generation and subsequent mating pool.\n",
        "\n",
        "What is possibly more interesting from this example is that as the generations progress and the population gets fitter the gene pattern for individual No. 1 will become more prevalent in the general population because it is fitter, more apt to the environment we have put it in.\n",
        "\n",
        "The selection can be performed in the following way:\n",
        "- Compute the fitness of each chromosome of the population $p$\n",
        "- Compute the \\% fintess of each chromosome w.r.t the total fitness\n",
        "- Find the commulative density of each chromosome (from first to the last)\n",
        "- Generate a random number between 1 and 100. Select the chromosome in which range of accumulative \\% fitness occurs the random number \n",
        "\n",
        "In the following function you have to implement the selection function.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "LMtn-wXKOJCn",
        "outputId": "98bf6647-2b62-413f-c8b0-e155c64a933c"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[1, 3, 2, 3]\n"
          ]
        }
      ],
      "source": [
        "def selection(p):\n",
        "    sum_fitness=0\n",
        "    pf=[]# percent fitness\n",
        "    for c in p:\n",
        "        sum_fitness+=c[0]\n",
        "    \n",
        "    cpf=0 #cummolative fitness\n",
        "    rn=rnd.randint(0,99) #Generate a random number between 0 and 100\n",
        "\n",
        "    for c in p:\n",
        "        cpf+=(c[0]*100)/sum_fitness\n",
        "        if rn<=cpf:\n",
        "            return c[1]\n",
        "    return p[rnd.randint(0,len(p)-1)]\n",
        "        \n",
        "print(selection(p))\n",
        "\n",
        "              "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bWt6I3yFOJCo"
      },
      "source": [
        "### crossover:\n",
        "A genetic operation by which new chromosomes are created from the current chromosomes.  Here, you have to use a single point crossover. A point on both parents' chromosomes is picked randomly, and designated a 'crossover point'. Bits to the right of that point are swapped between the two parent chromosomes. This results in two offspring, each carrying some genetic information from both parents. The following figure shows the a single point crossover operation.\n",
        "\n",
        "<img src=\"OnePointCrossover.svg.png\">\n",
        "<caption><center> <u> <font color='purple'> **Single point crossover operation** </u></center></caption>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_FPzJmTqOJCo"
      },
      "outputs": [],
      "source": [
        "def crossover(x,y):\n",
        "    # find the spliting point sp: randomly generate a nuumber between 1 and chromosome_length-2\n",
        "\n",
        "    sp=None\n",
        "  \n",
        "    nx=[] # new child 1\n",
        "    ny=[] # new child 2\n",
        "    nx=None # new child 1, which should contain [0,sp] genes from x and [sp, chromosome_length] genes from y\n",
        "    ny=None# new child 2, which should contain [0,sp] genes from y and [sp, chromosome_length] genes from x\n",
        "    return (nx,ny)\n",
        "    "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "LsknX_E0OJCp",
        "outputId": "9fd015ad-2663-4576-faad-a158d24afb9a"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[0, 2, 2, 0]\n",
            "[3, 3, 1, 2]\n",
            "[0, 3, 1, 2] [3, 2, 2, 0]\n"
          ]
        }
      ],
      "source": [
        "print(p[1][1])\n",
        "print(p[2][1])\n",
        "x,y=crossover(p[1][1],p[2][1])\n",
        "print(x,y)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "SE9SpKaAOJCp",
        "outputId": "b048a021-9061-438a-c6cd-72d25c223d4f"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "0.688939048363144\n"
          ]
        }
      ],
      "source": [
        "# In the mutation operation you will need a random number between 0 and 1. \n",
        "# The following function is used to generate the required number.\n",
        "from random import random\n",
        "print(random())\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0qZG8uMOOJCp"
      },
      "source": [
        "### Mutation:\n",
        "Mutation is a genetic operator used to maintain genetic diversity from one generation of a population of genetic algorithm chromosomes to the next. It is analogous to biological mutation. Mutation alters one or more gene values in a chromosome from its initial state. In mutation, the solution may change entirely from the previous solution. Hence GA can come to a better solution by using mutation. Mutation occurs during evolution according to a user-definable mutation probability. This probability should be set low. If it is set too high, the search will turn into a primitive random search.\n",
        "\n",
        "The classic example of a mutation operator involves a probability that an arbitrary bit in a genetic sequence will be changed from its original state. A common method of implementing the mutation operator involves generating a random variable for each bit in a sequence. This random variable tells whether or not a particular bit will be modified. This mutation procedure, based on the biological point mutation, is called single point mutation. Other types are inversion and floating point mutation. When the gene encoding is restrictive as in permutation problems, mutations are swaps, inversions, and scrambles.\n",
        "\n",
        "Here, you have to do a single point mutation as follows:\n",
        "- generate a random number $rn$\n",
        "- if the $rn\\le mp$ then change any one of the gene value. Here, $mp$, is the mutation probability. \n",
        "- Make sure that the new value of the gene lies in the range $[gene_{min}, gene_{max}]$"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "qv_ZG5rTOJCr"
      },
      "outputs": [],
      "source": [
        "def mutation(x,mp,gene_min,gene_max):\n",
        "    rn=None # generate a random number [0,1]\n",
        "    if rn<=mp:\n",
        "        x[None]=None # Change the given gene value with a new gene value. Note that the new value should be in range [gene_min,gene_max]\n",
        "    return x\n",
        "        "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "512hzKM2OJCr",
        "outputId": "1df057d0-6847-4190-a178-761a2a7b405a"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[4, 2, 2, 0]\n"
          ]
        }
      ],
      "source": [
        "gene_min=1\n",
        "gene_max=4\n",
        "mp=0.9\n",
        "print(mutation(p[1][1],mp,gene_min,gene_max))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RBUyIOHZOJCs"
      },
      "source": [
        "<img src=\"algo.png\">\n",
        "<caption><center> <u> <font color='purple'> **A pseudocode of genetic algorithm** </u></center></caption>\n",
        "    "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "qUzln7fyOJCs"
      },
      "outputs": [],
      "source": [
        "def GA(p_size,ch_length,gene_min,gene_max,mp,epochs,stoping_condtion):\n",
        "    # The arguments p_size,ch_length,gene_min,gene_max,mp, represents the\n",
        "    # population size, chromosome length,minimum value of gene, maximum value of gene, \n",
        "    # mutation probability, the maximum number of epochs respectively\n",
        "    \n",
        "    p=None #Generate the initial population p\n",
        "    \n",
        "    #compute the fitness of the population p\n",
        "    #start code\n",
        "    None\n",
        "    # end code\n",
        "    \n",
        "    stop=0 #This is the stoping criteria.\n",
        "    gen=1\n",
        "    while(gen<epochs):\n",
        "        pp=p # The previous population\n",
        "        # Generate the new population\n",
        "        np=[] # Initialize the new population\n",
        "        for i in range(p_size):\n",
        "            x=None # select the first chromosome for crossover\n",
        "            y=None # select the second chromosome for crossover\n",
        "            # Perform crossover\n",
        "            x,y=None #perform crossover\n",
        "            #perform mutaion\n",
        "            x=None # Mutate x\n",
        "            y=None # mutate y\n",
        "            # add the chromosome x, and y to the new population\n",
        "            np.append(x)\n",
        "            np.append(y)\n",
        "            # Replace the old population p with new population np\n",
        "        p=None\n",
        "        #print(p)\n",
        "        #Compute the fitness of the population p\n",
        "        pf=[]\n",
        "        for c in p:\n",
        "            pf.append([fitness(c),c])\n",
        "        p=pf\n",
        "        \n",
        "        #print(p)\n",
        "        # sort the current population p and previous population pp on the basis of fitness \n",
        "        pp.sort()\n",
        "        p.sort()\n",
        "        # Perform Elitisim\n",
        "        #p[0]=pp[p_size-1]\n",
        "        if pp[p_size-1][0]== p[p_size-1][0]:\n",
        "            stop+=1\n",
        "        else:\n",
        "            stop=0\n",
        "        if stop==stoping_condtion or p[p_size-1][0]==28:\n",
        "            return p[p_size-1]\n",
        "        gen+=1\n",
        "            \n",
        "    return p[p_size-1]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "6VP_Nz9nOJCt"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "EpaGQfS0OJCt",
        "outputId": "5043eddd-68ff-4058-dc1c-69195776a34b"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "The solution is [20.0, [2, 6, 8, 6, 6, 1, 1, 8]]\n"
          ]
        }
      ],
      "source": [
        "p_size=20\n",
        "ch_length=8\n",
        "gene_min=1\n",
        "gene_max=8\n",
        "mp=0.02\n",
        "\n",
        "epochs=10000000\n",
        "stoping_condtion=3\n",
        "\n",
        "solution=GA(p_size,ch_length,gene_min,gene_max,mp,epochs,stoping_condtion)\n",
        "\n",
        "print(\"The solution is\", solution)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "S58M9EqIOJCt"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hzbRVCqcOJCu"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "054BEg1aOJCu"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "YVrTtyVnOJCu"
      },
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.6.5"
    },
    "colab": {
      "provenance": []
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}